### ISSUE TITLE
Implement Direct Sockets API

### ISSUE THREAD

**Issue by meponderR** [emscripten-core/emscripten#16737]
The [Direct Sockets API](https://github.com/WICG/direct-sockets/blob/main/docs/explainer.md) is currently in trial in `chrome://flags`. If this were to be implemented it would allow network sockets to work within Emscripten in a web browser without the need of a proxy.

**Comment by sbc100** [comment]
That does look promising.  Looks like it should be fairly simple to hook up to our posix socket layer if anyone wants to have a go.

**Comment by guest271314** [comment]
Here's my implementation of `TCPSocket` and `UDPSocket` using IPC via Web extension `external_connecatable`, written to WHATWG Streams https://github.com/guest271314/sockets/blob/main/sockets-web-extension/direct-socket.js.

**Comment by LevitatingBusinessMan** [comment]
might be good to note that firefox took a negative position: https://github.com/mozilla/standards-positions/issues/431

**Comment by guest271314** [comment]
> might be good to note that firefox took a negative position: [mozilla/standards-positions#431](https://github.com/mozilla/standards-positions/issues/431)

Yeah, Firefox doesn't do a lot of stuff that's capable in Chromium-based browsers.

**Comment by guest271314** [comment]
> might be good to note that firefox took a negative position: [mozilla/standards-positions#431](https://github.com/mozilla/standards-positions/issues/431)

Keep in mind it's possible to create TCP, UDP, and any other networking connections, and send data over those connections from and to the browser using a Web extension and Native Messaging.
